lib/graphql/query/serial_execution/field_resolution.rb

```diff
@@ -32,7 +32,10 @@ def result
         # continue by "finishing" the value, eg. executing sub-fields or coercing values
         def get_finished_value(raw_value)
           raise raw_value if raw_value.instance_of?(GraphQL::ExecutionError)
-          return nil if raw_value.nil?
+          if raw_value.nil?
+            raise GraphQL::ExecutionError, "Type mismatch resolving '#{field.name}'" if field.type.kind.non_null?
```

Comment:
I think this should be done in GraphQL::NonNullType#coerce_result if `of_type.coerce_result(value)` returns `nil`.  Doing it here means we are still missing the check on elements of a list.